src/agent/rustjail/src/cgroups/fs/mod.rs

```diff
@@ -27,1120 +43,848 @@ macro_rules! sl {
+        DevicePermissions::MkNod,
+    ];
 
-    let v = format!(
-        "{} {}:{} {}",
-        t,
-        major.as_str(),
-        minor.as_str(),
-        d.access.as_str()
-    );
-
-    info!(sl!(), "{}", v.as_str());
-
-    write_file(dir, file, v.as_str())
-}
-
-impl Subsystem for Devices {
-    fn name(&self) -> String {
-        "devices".to_string()
+    DeviceResource {
+        allow: true,
+        devtype: dev_type,
+        major: d.major,
+        minor: d.minor,
+        access: permissions,
     }
+}
 
-    fn set(&self, dir: &str, r: &LinuxResources, _update: bool) -> Result<()> {
-        for d in r.devices.iter() {
-            write_device(d, dir)?;
-        }
-
-        for d in DEFAULT_DEVICES.iter() {
-            let td = LinuxDeviceCgroup {
-                allow: true,
-                r#type: d.r#type.clone(),
-                major: Some(d.major),
-                minor: Some(d.minor),
-                access: "rwm".to_string(),
-            };
-
-            write_device(&td, dir)?;
-        }
+fn linux_device_group_to_cgroup_device(d: &LinuxDeviceCgroup) -> DeviceResource {
+    let dev_type = string_to_device_type(&d.r#type);
 
-        for d in DEFAULT_ALLOWED_DEVICES.iter() {
-            write_device(d, dir)?;
+    let mut permissions: Vec<DevicePermissions> = vec![];
+    for p in d.access.chars().collect::<Vec<char>>() {
+        match p {
+            'r' => permissions.push(DevicePermissions::Read),
+            'w' => permissions.push(DevicePermissions::Write),
+            'm' => permissions.push(DevicePermissions::MkNod),
+            _ => {}
         }
-
-        Ok(())
     }
-}
 
-fn try_write<T>(dir: &str, file: &str, v: T) -> Result<()>
-where
-    T: ToString,
-{
-    match write_file(dir, file, v) {
-        Err(ref e)
-            if io_error_kind_eq(e, std::io::ErrorKind::PermissionDenied)
-                || io_error_kind_eq(e, std::io::ErrorKind::Other) =>
-        {
-            return Ok(());
-        }
-
-        Err(e) => return Err(anyhow!(e)),
-
-        Ok(_) => return Ok(()),
+    DeviceResource {
+        allow: d.allow,
+        devtype: dev_type,
+        major: d.major.unwrap_or(0),
+        minor: d.minor.unwrap_or(0),
+        access: permissions,
     }
 }
 
-impl Subsystem for Memory {
-    fn name(&self) -> String {
-        "memory".to_string()
-    }
-
-    fn set(&self, dir: &str, r: &LinuxResources, update: bool) -> Result<()> {
-        if r.memory.is_none() {
-            return Ok(());
-        }
-
-        let memory = r.memory.as_ref().unwrap();
-        // initialize kmem limits for accounting
-        if !update {
-            try_write(dir, KMEM_LIMIT, 1)?;
-            try_write(dir, KMEM_LIMIT, -1)?;
+fn line_to_vec(line: &str) -> Vec<u64> {
+    let mut m = Vec::new();
+    for n in line.split(' ') {
+        if !n.trim().is_empty() {
+            match n.trim().parse::<u64>() {
+                Ok(v) => m.push(v),
+                Err(_) => {}
```

Comment:
OK. Not very harmful, but weird.

Reply:
Yes, the functional aspect of Rust is cool